File: listing_builder/frontend/src/lib/api/alertSettings.ts
// frontend/src/lib/api/alertSettings.ts
// Purpose: API calls for Sellerboard-style alert settings
// NOT for: React hooks or UI components

import { apiRequest } from './client'

export interface AlertTypeInfo {
  type: string
  label: string
  category: string
  data_source: string
  default_priority: string
  active: boolean
}

export interface AlertSettingResponse extends AlertTypeInfo {
  alert_type: string
  priority: string
  notify_in_app: boolean
  notify_email: boolean
  email_recipients: string[]
  enabled: boolean
}

export interface AlertTypeSettingPayload {
  alert_type: string
  priority: string
  notify_in_app: boolean
  notify_email: boolean
  email_recipients: string[]
  enabled: boolean
}

export async function fetchAlertTypes(): Promise<AlertTypeInfo[]> {
  const response = await apiRequest<AlertTypeInfo[]>('get', '/alert-settings/types')
  if (response.error) throw new Error(response.error)
  return response.data!
}

export async function fetchAlertSettings(): Promise<AlertSettingResponse[]> {
  const response = await apiRequest<AlertSettingResponse[]>('get', '/alert-settings')
  if (response.error) throw new Error(response.error)
  return response.data!
}

export async function updateAlertSettings(
  settings: AlertTypeSettingPayload[]
): Promise<{ status: string; count: number }> {
  const response = await apiRequest<{ status: string; count: number }>(
    'put', '/alert-settings', { settings }
  )
  if (response.error) throw new Error(response.error)
  return response.data!
}

export async function toggleAlertSetting(
  alertType: string
): Promise<{ alert_type: string; enabled: boolean }> {
  const response = await apiRequest<{ alert_type: string; enabled: boolean }>(
    'patch', `/alert-settings/${alertType}`
  )
  if (response.error) throw new Error(response.error)
  return response.data!
}

File: listing_builder/backend/schemas/alert_settings.py
# backend/schemas/alert_settings.py
# Purpose: Pydantic schemas for Sellerboard-style alert settings
# NOT for: ORM models, registry data, or route logic

from pydantic import BaseModel, Field
from typing import List


class AlertTypeInfo(BaseModel):
    """Registry entry returned by GET /alert-settings/types."""
    type: str
    label: str
    category: str
    data_source: str
    default_priority: str
    active: bool  # WHY: True if the alert works now (no SP-API needed)


class AlertTypeSetting(BaseModel):
    """Single alert type setting as saved by user."""
    alert_type: str = Field(..., max_length=80)
    priority: str = Field(default="minor", pattern="^(minor|major|critical)$")
    notify_in_app: bool = True
    notify_email: bool = False
    email_recipients: List[str] = Field(default=[], max_length=10)
    enabled: bool = True


class AlertSettingsBulkUpdate(BaseModel):
    """Bulk upsert payload from the frontend Save button."""
    # WHY max 50: Registry has 24 types — 50 is generous ceiling to prevent abuse
    settings: List[AlertTypeSetting] = Field(..., max_length=50)


class AlertSettingResponse(AlertTypeSetting):
    """What the user gets back — setting merged with registry info."""
    label: str
    category: str
    data_source: str
    active: bool

    class Config:
        from_attributes = True
